### PR TITLE
Grafana dashboard grafana 5 compatible

### LINKED ISSUE #39663: grafana_dashboard fails with Grafana v5.0
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
grafana_dashboard

##### ANSIBLE VERSION
```
ansible 2.5.0
  config file = /Users/remirey/dev/grtgas/avignon/ansible.cfg
  configured module search path = [u'/Users/remirey/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/remirey/.virtualenvs/avignon/lib/python2.7/site-packages/ansible
  executable location = /Users/remirey/.virtualenvs/avignon/bin/ansible
  python version = 2.7.13 (default, Jul 18 2017, 09:17:00) [GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.42)]
```

##### CONFIGURATION
```
DEFAULT_GATHERING(/Users/remirey/dev/grtgas/avignon/ansible.cfg) = explicit
DEFAULT_HASH_BEHAVIOUR(/Users/remirey/dev/grtgas/avignon/ansible.cfg) = replace
DEFAULT_MANAGED_STR(/Users/remirey/dev/grtgas/avignon/ansible.cfg) = Please do not change this file directly since it is managed by Ansible and will be overwritten
DEFAULT_VAULT_PASSWORD_FILE(/Users/remirey/dev/grtgas/avignon/ansible.cfg) = /Users/remirey/dev/grtgas/avignon/ansible_vault_password
HOST_KEY_CHECKING(/Users/remirey/dev/grtgas/avignon/ansible.cfg) = False
RETRY_FILES_ENABLED(/Users/remirey/dev/grtgas/avignon/ansible.cfg) = False
```

##### OS / ENVIRONMENT
N/A

##### SUMMARY

When executing a dashboard creation to test idempotency, the second execution fails with the following error:

```
The full traceback is:
  File "/var/folders/nx/br96pdj13gd30xmz81_fpsw00000gn/T/ansible_NgOhbA/ansible_module_grafana_dashboard.py", line 307, in main
    result = grafana_create_dashboard(module, module.params)
  File "/var/folders/nx/br96pdj13gd30xmz81_fpsw00000gn/T/ansible_NgOhbA/ansible_module_grafana_dashboard.py", line 215, in grafana_create_dashboard
    raise GrafanaAPIException('Unable to create the new dashboard %s : %s - %s.' % (slug, info['status'], info))

failed: [localhost] (item={u'path': u'aws_ecs.json', u'slug': u'aws_ecs'}) => {
    "changed": false,
    "invocation": {
        "module_args": {
            "grafana_api_key": null,
            "grafana_password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "grafana_url": "<some_url>",
            "grafana_user": "<some_user>",
            "message": "Updated by ansible",
            "org_id": 1,
            "overwrite": true,
            "path": "files/grafana/dashboards/aws_ecs.json",
            "slug": "aws_ecs",
            "state": "present",
            "validate_certs": true
        }
    },
    "item": {
        "path": "aws_ecs.json",
        "slug": "aws_ecs"
    },
    "msg": "error : Unable to create the new dashboard aws_ecs : 412 - {'body': '{\"message\":\"The dashboard has been changed by someone else\",\"status\":\"version-mismatch\"}', 'date': 'Thu, 03 May 2018 09:37:05 GMT', 'content-length': '88', 'status': 412, 'url': '<some_url>/api/dashboards/db', Expires=Thu, 10 May 2018 09:37:05 GMT; Path=/', 'content-type': 'application/json', 'connection': 'close', 'msg': 'HTTP Error 412: Precondition Failed'}."
```

##### STEPS TO REPRODUCE

```
    - name: Import the Grafana dashboards
      grafana_dashboard:
        grafana_url: <some_url>
        grafana_user: "{{ grafana_admin_user }}"
        grafana_password: "{{ grafana_admin_password }}"
        state: present
        message: Updated by ansible
        overwrite: yes
        path: "files/grafana/dashboards/{{ item.path }}"
        slug: "{{ item.slug }}"
      with_items:
        - path: aws_ecs.json
          slug: aws_ecs
```

The meta section was added manually in the dashboard json to have the import work in the first place.

##### EXPECTED RESULTS
Idempotency works and second execution marks the task "ok" and not "changed".

##### ACTUAL RESULTS
Exception

```
The full traceback is:
  File "/var/folders/nx/br96pdj13gd30xmz81_fpsw00000gn/T/ansible_NgOhbA/ansible_module_grafana_dashboard.py", line 307, in main
    result = grafana_create_dashboard(module, module.params)
  File "/var/folders/nx/br96pdj13gd30xmz81_fpsw00000gn/T/ansible_NgOhbA/ansible_module_grafana_dashboard.py", line 215, in grafana_create_dashboard
    raise GrafanaAPIException('Unable to create the new dashboard %s : %s - %s.' % (slug, info['status'], info))
```

### PR DESCRIPTION
**Pull request by seuf**
##### SUMMARY
Update grafana dashboard to be compatible with grafana v5+
Fix #39663

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
grafana_dashboard module

##### ANSIBLE VERSION
```
ansible 2.5.3
  config file = None
  configured module search path = [u'/home/seuf/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/seuf/python-virtualenv-ansible/lib/python2.7/site-packages/ansible
  executable location = /home/seuf/python-virtualenv-ansible/bin/ansible
  python version = 2.7.15 (default, May  2 2018, 14:12:52) [GCC 8.0.1 20180324 (Red Hat 8.0.1-0.20)]
```

##### ADDITIONAL INFORMATION
This PR is retro compatible with grafana 4.

### CODE REVIEW

**Comment by samdoran** [comment]
Thanks for the fix for this module! The Grafana 5 support looks good.

However, the addition of the `list` functionality in a module goes against the design of modules, which is to achieve a desired state. Since `list` isn't really a valid state, this capability is better suited to a [lookup plugin](https://docs.ansible.com/ansible/devel/dev_guide/developing_plugins.html#lookup-plugins) or facts module.

File: lib/ansible/modules/monitoring/grafana_dashboard.py [hunk 1]
```diff
@@ -47,10 +47,15 @@
     description:
       - State of the dashboard.
     required: true
-    choices: [ absent, export, present ]
+    choices: [ absent, export, present, list ]
     default: present
+  uid:
```
Comment (achinthagunasekara) [inline lib/ansible/modules/monitoring/grafana_dashboard.py]: Perhaps move this after `slug` to preservice alphabetical ordering?

Comment (seuf) [reply lib/ansible/modules/monitoring/grafana_dashboard.py]: ok, i've moved uid below slug.

**Comment by seuf** [comment]
@samdoran OK, I've removed the `list` value for `state` param.
I will create a new pull request for a grafana dashboard lookup plugin later.

File: lib/ansible/modules/monitoring/grafana_dashboard.py [hunk 2]
```diff
@@ -125,26 +134,63 @@ class GrafanaExportException(Exception):
...
 
 
-def grafana_dashboard_exists(module, grafana_url, slug, headers):
+def grafana_headers(module, data):
+    headers = {'content-type': 'application/json; charset=utf8'}
+    if 'grafana_api_key' in data and data['grafana_api_key']:
+        headers['Authorization'] = "Bearer %s" % data['grafana_api_key']
+    else:
+        auth = base64.b64encode(to_bytes('%s:%s' % (data['grafana_user'], data['grafana_password'])).replace('\n', ''))
+        headers['Authorization'] = 'Basic %s' % auth
```
Comment (rrey) [inline lib/ansible/modules/monitoring/grafana_dashboard.py]: Is there a particular reason for not using the fetch_url mechanism for the basic auth ?

File: lib/ansible/modules/monitoring/grafana_dashboard.py [hunk 3]
```diff
@@ -288,6 +364,7 @@ def main():
             grafana_password=dict(default='admin', no_log=True),
             grafana_api_key=dict(type='str', no_log=True),
             org_id=dict(default=1, type='int'),
+            uid=dict(type='str'),
```
Comment (rrey) [inline lib/ansible/modules/monitoring/grafana_dashboard.py]: shouldn't did and slug be added in the mutually_exclusive list ?

File: lib/ansible/modules/monitoring/grafana_dashboard.py (hunk 2 above)
Comment (seuf) [reply lib/ansible/modules/monitoring/grafana_dashboard.py]: Mhh.. I didn't know that the `fetch_url` was reading the `url_username` and `url_password` from module params.
I will fix it.

File: lib/ansible/modules/monitoring/grafana_dashboard.py (hunk 3 above)
Comment (seuf) [reply lib/ansible/modules/monitoring/grafana_dashboard.py]: Yep, good point !

File: lib/ansible/modules/monitoring/grafana_dashboard.py [hunk 4]
```diff
@@ -125,26 +132,64 @@ class GrafanaExportException(Exception):
...
         raise GrafanaAPIException('Unable to switch to organization %s : %s' % (org_id, info))
 
 
-def grafana_dashboard_exists(module, grafana_url, slug, headers):
+def grafana_headers(module, data):
+    headers = {'content-type': 'application/json; charset=utf8'}
+    if 'grafana_api_key' in data and data['grafana_api_key']:
+        headers['Authorization'] = "Bearer %s" % data['grafana_api_key']
+    else:
+        module.params['url_username'] = data.get('grafana_user')
```
Comment (rrey) [inline lib/ansible/modules/monitoring/grafana_dashboard.py]: You should check the url_argument_spec function in the urls helpers.
It return an argument spec that contains all the arguments that fetch_url supports (including url_username etc ...)
You can still override the values to add grafana_user and the previous vars as alias.

**Review by rrey — CHANGES_REQUESTED** [review]
Sorry for the late feedback ..

File: lib/ansible/modules/monitoring/grafana_dashboard.py (hunk 4 above)
Comment (seuf) [reply lib/ansible/modules/monitoring/grafana_dashboard.py]: Hi, Do you mean I should use the url_argument_spec in the main function ? then update tha argument_spec var with all the grafana params ? Ho do I create an alias for url_username <-> grafana_user ?

Comment (rrey) [reply lib/ansible/modules/monitoring/grafana_dashboard.py]: > Do you mean I should use the url_argument_spec in the main function ? then update tha argument_spec var with all the grafana params ?

yes, I saw that in other modules like `web_infrastructure/jenkins_plugin.py` (line 715)

>  Ho do I create an alias for url_username <-> grafana_user ?

In the argument_spec, in the dict you have the `aliases`key. 
Check the file module `files/file.py` line 181.
The `path` parameter has aliases `dest` and `name` for backward compatibility.

**Comment by seuf** [comment]
Hi @rrey ! I've found time to fix this module :) 
As you asked I've used the `url_argument_spec` function to init modules params.

**Comment by pierrefevrier** [comment]
Hi @seuf !  
Thanks for your PR, I have the same issue with Grafana 5.1.3.  
I'm not sure `uid` parameter instead of `slug` one is the best option because we don't have an `uid` at the dashboard creation time.  
Instead I think it is possible to use the dashboard `title` to search for an existing dashboard (`http://grafana-url/api/search?query=<dashboard-title>`) and get the `uid` if exists, else create a new dashboard.  
This way it would be idempotent.

**Comment by seuf** [comment]
The `uid` argument is not mandatory when you import a new dashboard.
If you don't specify an uid argument, the module will check if an uid is present in your dashboard json file, if yes, it checks for an existing dashboard with the same uid to update it (if overwrite is true), else a new dashboard is created.

**Comment by rrey** [comment]
lgtm

**Comment by pierrefevrier** [comment]
I think the right thing to do is something "upsertable" (run the same command to get the same result), else how can I make this tool devopsable ?  

My workflow is the following:  I deploy Grafana dashboard on several envs (DEV, Q/A, PPROD, PROD). The first time I deploy the board, I don't know the `uid`, and I don't want to hard code the `uid` after the first deployment for each env.

**Comment by rrey** [comment]
Hi @pierrefevrier,

What you describe is fulfilled by the `uuid` use. When you create a dashboard in Grafana an `uuid` is automatically created and you can see it in the dashboard json. You don't have to bother about its value.

As soon as you export the json to deploy it on other environment, you are ready to go.
The module configuration will not need the `uuid`parameter to be provided as a value is already in the json.
Additionally, the `uuid` also aims at identifying the dashboard on multiple instance of Grafana, meaning the same `uuid` for the same dashboard in all your instance (dev, Q/A, preprod and prod).

The `uuid` parameter in the module can allow you to override the value in the json, I doubt a bit of its usefulness but maybe @seuf has a use case in mind.

Concerning the use of the title that you suggest, IMHO the title is something that you may want to change often for several reason (typos, no more reflecting content, bad name in the first place, etc ...).
Ansible will not rename the previous dashboard deployment, it will create a new one and leave the old one. (because there are no state file like in terraform tracking the resource id).
Doing clean deployments of a renamed dashboard will force you to add the deletion tasks in the playbook.

**Comment by pierrefevrier** [comment]
Hi @rrey,  
Thanks for your answer.  
Ok, I forgot to talk about a detail: I deploy the same dashboard's template for several envs on the same Grafana instance, so I need to use a different `uid` for each :/

**Comment by seuf** [comment]
Hi @pierrefevrier 
Since Grafana 5, the dashbord title is not unique. Multiple dashboards can have the same title with differents uid. You can just remove the uid from your json file and Grafana will create new dashboard each time.

**Comment by pierrefevrier** [comment]
@seuf: doing so will re-create the dashboard each time I deploy my app :S

**Comment by rrey** [comment]
If I'm not wrong, the feature allowing to erase the uuid solves your problem @pierrefevrier.
If you force the uid for your dashboard, you can instanciate it several times.

```
dev -> uid = dev_xxxx
stg -> uid = stg_xxxx
prod -> uid = prod_xxxx
```

@pierrefevrier, how do you change the datasource for each dashboard ?

**Comment by seuf** [comment]
@pierrefevrier Nope !
The grafana dashboard module compare existing dashboard with the same uid with the new json, then import it only if the `overwrite` parameter is set to yes !

See this part of code :
```
dashboard_exists, dashboard = grafana_dashboard_exists(module, data['grafana_url'], uid, headers=headers)

    result = {}
    if dashboard_exists is True:
        if dashboard == payload:
            # unchanged
            result['uid'] = uid
            result['msg'] = "Dashboard %s unchanged." % uid
            result['changed'] = False
        else:
            # update
            if 'overwrite' in data and data['overwrite'] == 'yes':
                payload['overwrite'] = True
            if 'message' in data and data['message']:
                payload['message'] = data['message']
```

**Comment by pierrefevrier** [comment]
@rrey thank's for your last answer, I didn't know it is possible to choose the `uid` of a new dashboard (after a test, I confirm we can choose it).  
About datasource, each dashboard has a different one.  
To make it work, I will use Ansible template from my dashboard's template to generate each dashboard instance.

I think my problem is solved right know (the only think I missed in fact was the possibility to choose the `uid`).

Thanks for your answers guys !

**Comment by seuf** [comment]
OK, so we are ready to merge !!! 🚀  cc @bcoca @rrey @samdoran

**Comment by samdoran** [comment]
@seuf Please create a changelog fragment. See [fragments](https://github.com/ansible/ansible/tree/stable-2.6/changelogs/fragments) for examples.

File: lib/ansible/modules/monitoring/grafana_dashboard.py [hunk 5]
```diff
@@ -318,10 +421,16 @@ def main():
             msg="error : json file does not contain a meta section with a slug parameter, or you did'nt specify the slug parameter"
         )
         return
+    except GrafanaDeleteException as e:
+        module.fail_json(
+            failed=True,
+            msg="error : Can't delete dashboard : %s" % str(e)
```
Comment (samdoran) [inline lib/ansible/modules/monitoring/grafana_dashboard.py]: Use `to_native()` for converting error messages to ensure messages are in the correct native string type for the Python version. See [Raising Errors](https://docs.ansible.com/ansible/latest/dev_guide/developing_plugins.html#raising-errors).

**Review by samdoran — CHANGES_REQUESTED** [review]

**Comment by seuf** [comment]
@samdoran Hi, I've updated the pr with `to_native` when raising errors (and rebase over devel). Can you take a look ?

**Comment by samdoran** [comment]
@seuf There are still a few uses of `str()`. Please change those as well.

**Comment by seuf** [comment]
Arf, I've missed it ! It should be ok now..

**Review by samdoran — APPROVED** [review]

**Comment by samdoran** [comment]
Please create a [backport PR](https://docs.ansible.com/ansible/latest/community/development_process.html#backport-pull-request-process) for this to be included in previous versions.

**Pull request merged**